### PR TITLE
New/swh/dynamic nav

### PR DESCRIPTION
**Pull request by SamWhite745**
- Created route for getting a username by email.
- Added dynamic nav-bar
    - Changes based on logged in context
    - also greets user when logged in by username
- Created Logged in context which can be imported and used around the program
- removed extra logout button from feed
- no tests cos im bad

### CODE REVIEW

File: api/controllers/users.js [hunk 1]
```diff
@@ -11,6 +11,15 @@ const UsersController = {
...
   },
+  GetUsername: (req, res) => {
+    const email = (req.params.email);
+    User.findOne({email: email}, async (err, user) => {
+      if (err) {
+        throw err;
+      }
+      res.status(200).json({ username: user.username, message: "Found user" });
+    });
+  }
```
Comment (siqbal181) [inline api/controllers/users.js:22]: Is this mainly to show "Welcome User" or for something else?

File: frontend/src/components/app/App.js [hunk 2]
```diff
@@ -1,35 +1,50 @@
...
+        <Route path='/posts'  element={
+          <ProtectedRoute>
+            <Feed navigate={ useNavigate() }/>
+          </ProtectedRoute>
+        }/>
+        <Route path='/login'  element={<LoginForm  navigate={ useNavigate() }/>}/>
+        <Route path='/signup' element={<SignUpForm navigate={ useNavigate() }/>}/>
+        <Route path='*' element={<Navigate to='/login' />} />
+      </Routes>
+    </div>
```
Comment (siqbal181) [inline frontend/src/components/app/App.js:46]: Hench code, very nice!

File: frontend/src/components/authenticationProvider/AuthenticationProvider.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import React, { createContext, useState } from 'react'
+
+export const AuthenticationContext = createContext({})
+
+const AuthenticationProvider = ({ children }) => {
+  const [isLoggedIn, setIsLoggedIn] = useState(false)
+  const [username, setUsername] = useState("")
+  return (
+    <AuthenticationContext.Provider value={{isLoggedIn, setIsLoggedIn, username, setUsername}}>
+      {children}
```
Comment (siqbal181) [inline frontend/src/components/authenticationProvider/AuthenticationProvider.js:11]: No idea how this works but assuming it is integral

File: frontend/src/components/feed/Feed.js [hunk 4]
```diff
@@ -25,20 +27,12 @@ const Feed = ({ navigate }) => {
   }, [needsRefresh]) //Dependency - when needsRefresh (a boolean)
     // changes, it call for the useEffect to be rerun - refreshing the posts
     // pass () => setRefresh(true) to any component that needs to refresh and call it in the component
-
-  const logout = () => {
-    window.localStorage.removeItem("token")
-    navigate('/login')
-  }
```
Comment (siqbal181) [inline frontend/src/components/feed/Feed.js:29]: Did the removeItem "toekn" on logout get moved somewhere else?

**Review by siqbal181 — APPROVED** [review]
Amazing work! Would be great to chat over as lot of this is going over my head code wise, but the context based on your PR comms makes sense :D

**Review by siqbal181 — DISMISSED** [review]
Amazing work! Would be great to chat over as lot of this is going over my head code wise, but the context based on your PR comms makes sense :D

**Review by siqbal181 — COMMENTED** [review]
Amazing code - updated note that testing on the side gives an error which I sent in slack

File: frontend/src/components/feed/Feed.js (hunk 4 above)
Comment (SamWhite745) [reply frontend/src/components/feed/Feed.js:29]: Whoops, new commit moves token functionality to the context too

File: api/controllers/users.js (hunk 1 above)
Comment (SamWhite745) [reply api/controllers/users.js:22]: Exactly :)

File: frontend/src/components/authenticationProvider/AuthenticationProvider.js (hunk 3 above)
Comment (SamWhite745) [reply frontend/src/components/authenticationProvider/AuthenticationProvider.js:11]: Yes this is how we make the 'context'. importing it in different components gives access to the hooks - essentially making them global - necessary for a dynamic navbar

File: frontend/src/index.js [hunk 5]
```diff
@@ -4,12 +4,15 @@ import './index.css';
...
 
 const root = ReactDOM.createRoot(document.getElementById('root'));
 root.render(
   <BrowserRouter>
     <React.StrictMode>
-      <App />
+      <AuthenticationProvider>
+        <App/>
+      </AuthenticationProvider>
     </React.StrictMode>
```
Comment (toppy007) [inline frontend/src/index.js:16]: looking forward to fiddling with this

**Review by toppy007 — DISMISSED** [review]
Wow amazing work! looking forward to seeing it in action.

**Review by toppy007 — APPROVED** [review]
all good

**Pull request merged**